### PR TITLE
Add changes to fix bad input from fuzzing, Fixes #210

### LINKED ISSUE #210: BadInputs cause unexpected behavior
**Description**
You can find bad inputs with corresponding stack trace in _unique_min_crashes_ and the byte input files in _CrashesDeduplicatedMinimized_:

[Run2 Results](https://github.com/JlKmn/nHapiMinCrashes/tree/master/Run2)

### PR DESCRIPTION
**Pull request by milkshakeuk**
Fixes #210

* Add new `SupportClass.CharsAreUnique` method
* Add check for unique encoding characters in `EncodingCharacters.cs` (was causing Stack Overflow)
* Add unit tests for changes plus additional Bad Input files to existing test
* Add additional validation for `EncodingCharacters.cs` to prevent whitespace (and Null) characters being used as encoding characters or as the field Seperator (MSH-1) + unit tests.

image

### CODE REVIEW

**Comment by milkshakeuk** [comment]
@JlKmn could you also have a quick review please

File: src/NHapi.Base/Parser/EncodingCharacters.cs [hunk 1]
```diff
@@ -52,13 +52,14 @@ public class EncodingCharacters : object, ICloneable
...
         /// </param>
+        /// <exception cref="HL7Exception">If encoding characters are not unique.</exception>
         public EncodingCharacters(char fieldSeparator, string encodingCharacters)
         {
             FieldSeparator = fieldSeparator;
 
             encChars = new char[4];
 
-            if (encodingCharacters == null)
+            if (string.IsNullOrEmpty(encodingCharacters))
```
Comment (AMCN41R) [inline src/NHapi.Base/Parser/EncodingCharacters.cs:62]: do we need to check whitespace?

File: src/NHapi.Base/SupportClass.cs [hunk 2]
```diff
@@ -2130,6 +2131,19 @@ public static object Pop(ArrayList stack)
...
         }
 
+        /// <summary>
+        /// Evaluates if a string is comprised of unique characters.
+        /// </summary>
+        /// <param name="input">string to evaluate.</param>
+        /// <returns>True if all characters are unique, otherwise False.</returns>
+        public static bool CharsAreUnique(string input)
+        {
+            var chars = input.ToCharArray();
```
Comment (AMCN41R) [inline src/NHapi.Base/SupportClass.cs:2141]: will `.ToCharArray()` throw if input is null?

File: tests/NHapi.Base.NUnit/SupportClassTests.cs [hunk 3]
```diff
@@ -0,0 +1,42 @@
+namespace NHapi.Base.NUnit
+{
+    using global::NUnit.Framework;
+
+    using NHapi.Base;
+
+    [TestFixture]
+    public class SupportClassTests
```
Comment (AMCN41R) [inline tests/NHapi.Base.NUnit/SupportClassTests.cs:8]: add a test for null/empty/whitespace input?

File: tests/NHapi.NUnit/Parser/EncodingCharactersTests.cs [hunk 4]
```diff
@@ -25,6 +25,35 @@ public void Constructor_ValidInput_ReturnsExpectedResult(char fieldSeperator, st
             Assert.AreEqual(encodingCharacters[3], sut.SubcomponentSeparator);
         }
 
+        [TestCase(null)]
+        [TestCase("")]
```
Comment (AMCN41R) [inline tests/NHapi.NUnit/Parser/EncodingCharactersTests.cs:29]: test for null?

File: src/NHapi.Base/Parser/EncodingCharacters.cs (hunk 1 above)
Comment (milkshakeuk) [reply src/NHapi.Base/Parser/EncodingCharacters.cs:62]: im not 100% sure but I would say no because they might want to use a whitespace character such as `\n` (this is an assumption). - what do you think?

File: src/NHapi.Base/SupportClass.cs (hunk 2 above)
Comment (milkshakeuk) [reply src/NHapi.Base/SupportClass.cs:2141]: most likely - will make some changes.

File: tests/NHapi.NUnit/Parser/EncodingCharactersTests.cs (hunk 4 above)
Comment (milkshakeuk) [reply tests/NHapi.NUnit/Parser/EncodingCharactersTests.cs:29]: not sure what you mean here?

File: tests/NHapi.Base.NUnit/SupportClassTests.cs (hunk 3 above)
Comment (milkshakeuk) [reply tests/NHapi.Base.NUnit/SupportClassTests.cs:8]: will do

File: tests/NHapi.NUnit/Parser/EncodingCharactersTests.cs (hunk 4 above)
Comment (AMCN41R) [reply tests/NHapi.NUnit/Parser/EncodingCharactersTests.cs:29]: sorry, I meant test for whitespace if you make the other change

File: src/NHapi.Base/Parser/EncodingCharacters.cs (hunk 1 above)
Comment (AMCN41R) [reply src/NHapi.Base/Parser/EncodingCharacters.cs:62]: yeah, could be valid - can leave for now and revisit if we need to?

File: tests/NHapi.Base.NUnit/SupportClassTests.cs (hunk 3 above)
Comment (milkshakeuk) [reply tests/NHapi.Base.NUnit/SupportClassTests.cs:8]: In regards to the generic `CharsAreUnique` whitespace chars are still valid chars - but will check for null or empty.

File: src/NHapi.Base/Parser/EncodingCharacters.cs (hunk 1 above)
Comment (milkshakeuk) [reply src/NHapi.Base/Parser/EncodingCharacters.cs]: image
Will perform whitespace checks and validations.

File: src/NHapi.Base/Parser/EncodingCharacters.cs [hunk 5]
```diff
@@ -52,13 +53,22 @@ public class EncodingCharacters : object, ICloneable
         /// Component Separator, Repetition Separator, Escape Character, and
         /// Subcomponent Separator (in that order).
         /// </param>
+        /// <exception cref="HL7Exception">If encoding characters are not unique.</exception>
         public EncodingCharacters(char fieldSeparator, string encodingCharacters)
         {
+            if (char.IsWhiteSpace(fieldSeparator) || fieldSeparator == char.MinValue)
+            {
+                throw new HL7Exception("Field Seperator must be a printable characters.");
```
Comment (AMCN41R) [inline src/NHapi.Base/Parser/EncodingCharacters.cs:61]: *character

File: src/NHapi.Base/Parser/EncodingCharacters.cs [hunk 6]
```diff
@@ -70,6 +80,16 @@ public EncodingCharacters(char fieldSeparator, string encodingCharacters)
             }
             else
             {
+                if (encodingCharacters.Any(@char => char.IsWhiteSpace(@char) || @char == char.MinValue))
+                {
+                    throw new HL7Exception("Encoding characters must be printable characters.");
```
Comment (AMCN41R) [inline src/NHapi.Base/Parser/EncodingCharacters.cs:85]: *character
